pkg/engine/api/engineresponse.go

```diff
@@ -199,7 +199,41 @@ func (er EngineResponse) GetValidationFailureAction() kyvernov1.ValidationFailur
 		return ""
 	}
 	spec := pol.AsKyvernoPolicy().GetSpec()
-	for _, v := range spec.GetValidationFailureActionOverrides() {
+	for _, r := range spec.Rules {
+		if r.HasValidate() {
+			for _, v := range r.Validation.ValidationFailureActionOverrides {
+				if !v.Action.IsValid() {
+					continue
+				}
+				if v.Namespaces == nil {
+					hasPass, err := utils.CheckSelector(v.NamespaceSelector, er.namespaceLabels)
+					if err == nil && hasPass {
+						return v.Action
+					}
+				}
+				for _, ns := range v.Namespaces {
+					if wildcard.Match(ns, er.PatchedResource.GetNamespace()) {
+						if v.NamespaceSelector == nil {
+							return v.Action
+						}
+						hasPass, err := utils.CheckSelector(v.NamespaceSelector, er.namespaceLabels)
+						if err == nil && hasPass {
+							return v.Action
+						}
+					}
+				}
+			}
+
+			if r.Validation.ValidationFailureAction != nil {
+				return *r.Validation.ValidationFailureAction
```

Comment:
We only return it for the first rule?

Reply:
Yes. This function is used by `BlockRequest()` to determine whether to block the request based on the ValidationFailureAction.

However, `BlockRequest()` is invoked by `HandleValidationEnforce()`, which applies policies with only `enforce` rules so it isn't necessary to check the action for each rule.

Reply:
BTW I'm considering removing this function but will create a follow-up PR.